runtime/lua/vim/iter.lua

```diff
@@ -1030,17 +1031,18 @@ function Iter.new(src, ...)
 
     local t = {}
 
-    -- O(n): scan the source table to decide if it is a list (consecutive integer indices 1…n).
-    local count = 0
-    for _ in pairs(src) do
-      count = count + 1
-      local v = src[count]
-      if v == nil then
+    -- O(n): scan the source table to decide if it is an array (only positive integer indices).
+    local length = 0
+    for k, v in pairs(src) do
+      if type(k) ~= 'number' or math.floor(k) ~= k then
         return Iter.new(pairs(src))
       end
-      t[count] = v
+      if k > length then
+        length = k
+      end
+      t[k] = v
     end
-    return ListIter.new(t)
+    return ArrayIter.new(t, length)
```

Comment:
IIUC, the `length` parameter is the main change. But it looks like we're also changing `t[k]` to use the (non-contiguous) number keys; will that cause any problems for `ArrayIter:next()`? 

Do we actually need the original number keys, or would it make sense to store `t` as a 1..n continguous "list" (as we did before)? Possibly I'm missing something, and maybe we need the number keys for `get()`-like functionality, or maybe we want to pass the number keys as `k` arg to the pipeline functions.

Reply:
> If we keep the non-contiguous keys in the new array, then we have the property that `vim.iter(t):totable() == t` (holes and all)

That seems desirable. OTOH, to preserve the non-contiguous keys the caller also has the option of passing `vim.iter(pairs(t))`, right?

Reply:
That's right. Another benefit of stripping holes is that any functions used in the iterator can behave consistently for both arrays and function iterators. A function iterator will never return `nil` (`nil` is the signal that the function iterator is done), so if we strip holes in arrays then we preserve that property.

```lua
vim.iter(s):filter(function(v)
  -- v can never be nil, regardless of the value of s
end)
```

I think it also makes the mental model more consistent. Right now if a pipeline stage returns `nil` (e.g. `:map()`) then that acts as removing an item from the iterator (future pipeline stages won't see that value). But if arrays were allowed to have nil values that would be inconsistent.

The more I think about it the more I do think that stripping `nil` is the right call, but we can always change it later if I'm wrong.

Reply:
As of now, the behavior is like the latter (i.e. `vim.iter(t)` strips holes, NOT `vim.iter(t):totable() == t`; e.g. `vim.iter({nil, 1, 2, nil, 3}):totable() == {1, 2, 3}`.

While I agree that what's proposed makes sense and has many benefits, I'm not 100% convinced whether this is the best choice, because: (1) `vim.iter(t):totable() == t` also sounds sensible; (2) what if one wants to keep `nil` for any reason (use `filter` explicitly to filter `nil`)?